torch/csrc/jit/passes/automatic_mixed_precision.cpp

```diff
@@ -0,0 +1,263 @@
+
+#include <torch/csrc/jit/passes/automatic_mixed_precision.h>
+
+#include <c10/core/ScalarType.h>
+#include <c10/util/Exception.h>
+#include <c10/util/Optional.h>
+#include <torch/csrc/jit/ir/ir.h>
+#include <torch/csrc/jit/jit_log.h>
+#include <torch/csrc/jit/passes/quantization/helper.h>
+
+#include <stack>
+#include <unordered_set>
+
+namespace torch {
+namespace jit {
+
+namespace {
+
+struct AutocastScope {
+  Value* instance = nullptr;
+  bool enabled = false;
+};
+
+// If we have an autocast instance, return it
+//
+// This is the pattern we're looking for (this is done after
+//  autocast.__init__() has been inlined)
+//
+// %4 : bool = prim::Constant[value=1]()
+// %5 : __torch__.torch.cuda.amp.autocast_mode.autocast = prim::CreateObject()
+//  = prim::SetAttr[name="_enabled"](%5, %4)
+//
+// Notes:
+//  1. There's no guarantee that the autocast instance is in the same block
+//    as the prim::Enter() node
+//  2. `prim::SetAttr` must follow `prim::CreateObject()` in the same block,
+//    but there might be other nodes in between
+//
+c10::optional<AutocastScope> parseAutocast(Value* value) {
+  if (value->node()->kind() == prim::CreateObject) {
+    const auto class_name = getModuleName(value);
+    if (class_name &&
+        *class_name == "__torch__.torch.cuda.amp.autocast_mode.autocast") {
+      // Search for `prim::SetAttr[name="_enabled"]`
+      for (Use use : value->uses()) {
+        if (use.user->kind() == prim::SetAttr &&
+            use.user->s(attr::name) == "_enabled") {
+          const auto def = use.user->input(1)->node();
+          if (def->kind() != prim::Constant) {
+            AT_ERROR("Autocast argument must be a constant");
+          }
+
+          // We have an autocast instance
+          AutocastScope scope;
+          scope.instance = value;
+          scope.enabled = def->i(attr::value) ? true : false;
+          return scope;
+        }
+      }
+    }
+  }
+
+  // Not an autocast...
+  return c10::nullopt;
+}
+
+void castTensorInputs(Node* node, at::ScalarType dtype) {
+  const auto graph = node->owningGraph();
+
+  WithInsertPoint insert_point(node);
+
+  const auto dtype_value = graph->insertConstant(dtype);
+  const auto false_value = graph->insertConstant(false);
+  const auto none_value = graph->insertConstant(IValue());
+
+  std::unordered_set<Value*> casted_inputs;
+
+  for (auto input : node->inputs()) {
+    if (input->type()->kind() == TensorType::Kind) {
+      casted_inputs.insert(input);
+    }
+  }
+
+  for (auto input : casted_inputs) {
+    const auto new_input = graph->insert(
+        aten::to, {input, dtype_value, false_value, false_value, none_value});
+    node->replaceInputWith(input, new_input);
+  }
+}
+
+void castInputsToWidestType(Node* node) {
+  // Figure out the widest type
+  // (really, just looking for any float32 inputs)
+  //
+  // TODO: revisit this (do we need to consider float64 types?)
```

Comment:
We can probably just error if AMP is on and we get a float64 tensor @mcarilli

Reply:
Interesting - we could add such a check. Anyway, the promote policy implementation is fairly broken as is right now and will require further changes - what we have here is just a stop gap to be able to run some real examples.